code/modules/dungeons/procedural/deepmaint.dm

```diff
@@ -0,0 +1,118 @@
+/*
+	For the sake of dungeon generator being modular and not tied exclusively to deepmaint,
+	most of the objects and modifications required exclusively for it will be kept here.
+*/
+
+var/global/list/free_deepmaint_ladders = list()
+/proc/check_deepmaint_list()
+	return (free_deepmaint_ladders.len)
+
+/obj/procedural/jp_DungeonGenerator/deepmaint
+	name = "Deep Maintenance Procedural Generator"
+
+
+/obj/procedural/jp_DungeonGenerator/deepmaint/generate(seed=null)
+	..(seed)
+	var/ladders_to_place = 4
+	if(numRooms < ladders_to_place)
+		return
+	var/list/obj/procedural/jp_DungeonRoom/done_rooms = list()
+	while(ladders_to_place > 0)
+		if(numRooms > 1)
+			if(done_rooms.len == out_rooms.len)
+				testing("Deepmaint generator went through all rooms, but couldn't place all ladders! Ladders left - [ladders_to_place]")
+				break
+		var/obj/procedural/jp_DungeonRoom/picked_room = pick(out_rooms)
+		if(picked_room in done_rooms)
+			continue
+		var/list/turf/viable_turfs = list()
+		for (var/turf/simulated/floor/F in range(roomMinSize + 1, picked_room.centre))
+			//not under walls
+			if (F.is_wall)
+				continue
+
+			//No airlocks
+			if (locate(/obj/machinery/door) in F)
+				continue
+
+			//No ladders or stairs
+			if (locate(/obj/structure/multiz) in F)
+				done_rooms += picked_room
+				viable_turfs = list()
+				break;
+
+			//No turfs in space
+			if (turf_is_external(F))
+				continue
+
+			//To be valid, the floor needs to have a wall in a cardinal direction
+			for (var/d in cardinal)
+				var/turf/T = get_step(F, d)
+				if (T.is_wall)
+					//Its got a wall!
+					viable_turfs[F] = T //We put this floor and its wall into the possible turfs list
```

Comment:
break out of the for loop here, you don't seem to even use the wall turf anyway.

Reply:
resolved